src/Symfony/Bundle/SecurityBundle/Command/DebugFirewallCommand.php

```diff
@@ -0,0 +1,178 @@
+ */
+
+namespace Symfony\Bundle\SecurityBundle\Command;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Bundle\SecurityBundle\Security\FirewallContext;
+use Symfony\Bundle\SecurityBundle\Security\LazyFirewallContext;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+/**
+ * @author Timo Bakx <timobakx@gmail.com>
+ */
+final class DebugFirewallCommand extends Command
+{
+    protected static $defaultName = 'debug:firewall';
+
+    /**
+     * @var string[]
+     */
+    private $firewallNames;
+
+    /**
+     * @var ContainerInterface
+     */
+    private $contexts;
+
+    public function __construct(array $firewallNames, ContainerInterface $contexts)
+    {
+        parent::__construct();
+
+        $this->firewallNames = $firewallNames;
+        $this->contexts = $contexts;
+    }
+
+    protected function configure(): void
+    {
+        $this
+            ->setDefinition([
+                new InputArgument('name', InputArgument::OPTIONAL, 'A firewall name (main)'),
+            ])
+        ;
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $io = new SymfonyStyle($input, $output);
+
+        $name = $input->getArgument('name');
+
+        if (null === $name) {
+            $this->displayFirewallList($io);
+
+            return 0;
+        }
+
+        $serviceId = sprintf('security.firewall.map.context.%s', $name);
+
+        if (!$this->contexts->has($serviceId)) {
+            $io->error(sprintf(
+                'Firewall %s was not found. Available firewalls are: %s',
+                $name,
+                implode(', ', $this->firewallNames)
+            ));
+
+            return 1;
+        }
+
+        /** @var FirewallContext $context */
+        $context = $this->contexts->get($serviceId);
+
+        $io->title(sprintf('Firewall "%s"', $name));
+
+        $this->displayFirewallSummary($context, $io);
+
+        $this->displaySwitchUser($context, $io);
+
+        $this->displayEventListeners($context, $io);
+
+        $this->displayAuthenticators($context, $io);
+
+        return 0;
+    }
+
+    protected function displayFirewallList(SymfonyStyle $io): void
+    {
+        $io->title('Firewalls');
+        $io->text('The following firewalls are defined:');
+
+        $rows = [];
+        foreach ($this->firewallNames as $name) {
+            $serviceId = sprintf('security.firewall.map.context.%s', $name);
+
+            if (!$this->contexts->has($serviceId)) {
+                continue;
+            }
+
+            $context = $this->contexts->get($serviceId);
+
+            if (null === $context->getConfig()) {
+                continue;
+            }
+
+            $rows[] = [
+                $context->getConfig()->getName(),
```

Comment:
Is there a difference between `$name` in the foreach loop and `$config->getName()` here? Or can we replace this foreach loop with `$rows = array_map('arrayval', $this->firewallNames)`?

Reply:
Ah, yes. Now we have the names array we could show that instead. I'll make that change.